### PR TITLE
refactor(receipts): Add a queue to perform work per conversation [WPB-9216]

### PR DESCRIPTION
**Pull request by vitorhugods**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9216" title="WPB-9216" target="_blank">BugWPB-9216</a>  [Android] out of memory exception when scrolling a group chat with long unread history
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

The app is _very_ slow to mark messages/conversations as read.

### Causes

1. We do stuff synchronously: send receipts, update lastReadDate, then send lastRead to other clients.
2. As the user scrolls through unread messages, we call the same `UpdateConversationReadDateUseCase` _multiple_ times in a row. And this use-case is _not_ prepared to only do one thing at a time. If called 30 times, it will launch 30 jobs in parallel for the same conversation. When only the "latest" one could/should be performed.

### Solutions

For this PR, I'm creating a structure that enables enqueuing work based on conversation ID and _only latest_ work will be performed.

If the idea of this PR is approved, I'll continue in next PRs with:

1. Allow fetching unread messages in a time window. Currently, DAOs are too opinionated and only allow fetching messages created _after_ the _lastRead_ – so, they require that _lastRead_ is only updated after sending receipts – which sucks.
2. Implementing a `ConversationReadTimeWorker`, which will do in parallel: gather messages/send receipts, update `lastRead` for the Conversation, send these receipts to the other devices of this user.

#### Test Coverage

- [X] I have added automated test to this contribution

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by MohamadJaara — APPROVED** [review]

**Review by yamilmedina — COMMENTED** [review]
Great work!  Some small suggestions added. I'm wondering if @alexandreferris  and I can use this since we are implementing `Delivery` receipts (even thought) these should be processed while receiving messages 🤔

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/receipt/ConversationTimeEventWork.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ */
+package com.wire.kalium.logic.feature.message.receipt
+
+import com.wire.kalium.logic.data.id.ConversationId
+import kotlinx.datetime.Instant
+
+internal data class ConversationTimeEventInput(
+    val conversationId: ConversationId,
+    val time: Instant,
+)
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/receipt/ConversationTimeEventWork.kt]: Done

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/receipt/ParallelConversationWorkQueue.kt [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+
+    private suspend fun createQueueForConversation(initialWork: ConversationTimeEventWork) =
+        MutableStateFlow(initialWork).also {
+            scope.launch(dispatcher) {
+                @Suppress("TooGenericExceptionCaught")
+                it.collect { work ->
+                    try {
+                        work.worker.doWork(work.conversationTimeEventInput)
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/receipt/ParallelConversationWorkQueue.kt]: Indeed, better!

**Comment by vitorhugods** [comment]
> Great work! Some small suggestions added. I'm wondering if @alexandreferris and I can use this since we are implementing `Delivery` receipts (even thought) these should be processed while receiving messages 🤔

It could be used for Delivery receipts. I had this in mind while working on it, actually. But I'm not sure if I covered everything.

I guess the `DeliveryReceiptsWorker` would be a bit more complex than the `ReadReceiptWorker`?
`DeliveryReceiptsWorker` can be quite sensitive during Sync. For instance, you probably need to wait until Sync is over, or you could risk sending receipts in conversations that no longer exist.

So either checking if the conversation still exists before actually starting to work, or not scheduling yet during sync, but rather keeping these in an auxiliary list and then scheduling _after_ sync is performed, or something else, not sure.

TL;DR: _Can_ be used, sure. Not sure of all the details.

**Review by yamilmedina — APPROVED** [review]
